From 76c351be623d8db3e45d7a0dd0a095cab75b3bc2 Mon Sep 17 00:00:00 2001
From: David Kershner <dkershner@habana.ai>
Date: Thu, 15 May 2025 21:57:50 +0300
Subject: [PATCH] [SW-210720] sensord: use set sensor threshold

Store user provided threshold information in boot_info to be used by
preboot for next hbm run.

Change-Id: Ie00b0001e38fc215aee9985edf16ebbf64d2a3e0
Signed-off-by: David Kershner <dkershner@habana.ai>
---

diff --git a/arch/gaudi3/ras/pldm/sensord_db.c b/arch/gaudi3/ras/pldm/sensord_db.c
index 96b0baf..e1b2eca 100644
--- a/arch/gaudi3/ras/pldm/sensord_db.c
+++ b/arch/gaudi3/ras/pldm/sensord_db.c
@@ -731,34 +731,42 @@
 	{
 		.pdr = &pldm_pdr_records[PLDM_RIDX_SENSOR_HBM_0_GLOBAL_ECC],
 		.api.get = sensord_get_hbm_global_ecc,
+		.api.set_sensor_thresholds = sensord_set_thresholds_hbm_global_ecc,
 	},
 	{
 		.pdr = &pldm_pdr_records[PLDM_RIDX_SENSOR_HBM_1_GLOBAL_ECC],
 		.api.get = sensord_get_hbm_global_ecc,
+		.api.set_sensor_thresholds = sensord_set_thresholds_hbm_global_ecc,
 	},
 	{
 		.pdr = &pldm_pdr_records[PLDM_RIDX_SENSOR_HBM_2_GLOBAL_ECC],
 		.api.get = sensord_get_hbm_global_ecc,
+		.api.set_sensor_thresholds = sensord_set_thresholds_hbm_global_ecc,
 	},
 	{
 		.pdr = &pldm_pdr_records[PLDM_RIDX_SENSOR_HBM_3_GLOBAL_ECC],
 		.api.get = sensord_get_hbm_global_ecc,
+		.api.set_sensor_thresholds = sensord_set_thresholds_hbm_global_ecc,
 	},
 	{
 		.pdr = &pldm_pdr_records[PLDM_RIDX_SENSOR_HBM_4_GLOBAL_ECC],
 		.api.get = sensord_get_hbm_global_ecc,
+		.api.set_sensor_thresholds = sensord_set_thresholds_hbm_global_ecc,
 	},
 	{
 		.pdr = &pldm_pdr_records[PLDM_RIDX_SENSOR_HBM_5_GLOBAL_ECC],
 		.api.get = sensord_get_hbm_global_ecc,
+		.api.set_sensor_thresholds = sensord_set_thresholds_hbm_global_ecc,
 	},
 	{
 		.pdr = &pldm_pdr_records[PLDM_RIDX_SENSOR_HBM_6_GLOBAL_ECC],
 		.api.get = sensord_get_hbm_global_ecc,
+		.api.set_sensor_thresholds = sensord_set_thresholds_hbm_global_ecc,
 	},
 	{
 		.pdr = &pldm_pdr_records[PLDM_RIDX_SENSOR_HBM_7_GLOBAL_ECC],
 		.api.get = sensord_get_hbm_global_ecc,
+		.api.set_sensor_thresholds = sensord_set_thresholds_hbm_global_ecc,
 	},
 	{
 		.pdr = &pldm_pdr_records[PLDM_RIDX_SENSOR_HBM_0_LANES_REPAIRED],
diff --git a/arch/gaudi3/ras/pldm/sensord_handlers.c b/arch/gaudi3/ras/pldm/sensord_handlers.c
index e0c58b5..8a5775e 100644
--- a/arch/gaudi3/ras/pldm/sensord_handlers.c
+++ b/arch/gaudi3/ras/pldm/sensord_handlers.c
@@ -1927,6 +1927,36 @@
 	return 0;
 }
 
+int sensord_set_thresholds_hbm_global_ecc(const struct sensord_sensor_entry *entry,
+					  struct pldm_pmc_threshold_data *thresholds)
+{
+	uint8_t idx = entry->pdr->common.record_handle - PLDM_RH_SENSOR_HBM_0_GLOBAL_ECC;
+	uint8_t hbm_thresholds[HL_NUM_HBM_COUNT];
+	int ret;
+
+	ret = hl_mem_info_rd_field(hbm_thresholds, hbm_ecc_thresholds);
+	if (ret) {
+		LOG_ERR("%s: failed to read threshold field: %d", __func__, ret);
+		return ret;
+	}
+
+	hbm_thresholds[idx] = MIN(0xFF, thresholds->upper_threshold_warning.u32);
+
+	ret = hl_mem_info_wr_field(hbm_thresholds, hbm_ecc_thresholds, false);
+	if (ret) {
+		LOG_ERR("%s: failed to read threshold field: %d", __func__, ret);
+		return ret;
+	}
+
+	ret = hl_boot_info_finalize();
+	if (ret) {
+		LOG_ERR("%s: Failed in boot_info ret=%d", __func__, ret);
+		return ret;
+	}
+
+	return 0;
+}
+
 int sensord_get_hbm_lanes_repaired(const struct sensord_sensor_entry *entry,
 				   struct sensord_sample *sample)
 {
diff --git a/include/arch/gaudi3/sensord_handlers.h b/include/arch/gaudi3/sensord_handlers.h
index f036550..79ba3bb 100644
--- a/include/arch/gaudi3/sensord_handlers.h
+++ b/include/arch/gaudi3/sensord_handlers.h
@@ -247,6 +247,8 @@
 				  struct sensord_sample *sample);
 int sensord_get_hbm_global_ecc(const struct sensord_sensor_entry *entry,
 			       struct sensord_sample *sample);
+int sensord_set_thresholds_hbm_global_ecc(const struct sensord_sensor_entry *entry,
+					  struct pldm_pmc_threshold_data *thresholds);
 int sensord_get_hbm_lanes_repaired(const struct sensord_sensor_entry *entry,
 				   struct sensord_sample *sample);
 int sensord_poll_hbm_utilization(const struct sensord_sensor_entry *entry,
